main.go

```diff
@@ -0,0 +1,28 @@
+package main
```

Comment:
> we could move all reusable protocol implementation code to a separate package (maybe protocol), and keep the server package for the reference implementation?

Yup.  From what's here, it seems like `server/message.go` and `server/serialization.go` can move to `protocol/` since I imagine they'd be used by both `client/` and `server/`.  But correct me if I'm wrong.

> Maybe I'm misunderstanding here. I thought that the server package is still a CONIKS library which implements the registration, key lookup functions... Then the developers can add some extra layers on the top of the library, e.g. storage backend which stores data that not in the tree, or something else ...

Maybe; then again, maybe not.  The storage backend just seems like a configuration though, not a reason for implementing your own layer.  Think about how, say MediaWiki, configure MySQL vs PostgreSQL.

> Or, they can choose to whether they want to use the registration bots library.

Hmm, I can see that working two ways:

 * If these are handled in process, the executable would pluggable (or, if we want to include them in the ref implementation, then configurable and off by default);
* But how I originally thought about these was that they'd be separate processes, and ther server would only accept registrations from the privileged places (ie. only registrations from localhost)

But ...

Please push back if the above isn't what you want.  The layering approach is also interesting and I can be convinced of it.